evm/vm/forks/sharding/shard_tracker.py

```diff
@@ -0,0 +1,161 @@
+        #       needed.
+        #       However, still need to figure out the window size. Since we need to make sure that
+        #       collation headers of non-best-candidate chains in the depth <= `WINDBACK_LENGTH`
+        #       should still be reachable.
+        self.headers_by_height = []
+
+    @to_tuple
+    def _get_new_logs(self):
+        shard_id_topic_hex = encode_hex(self.shard_id.to_bytes(32, byteorder='big'))
+        new_logs = self.log_handler.get_new_logs(
+            address=self.vmc_address,
+            topics=[
+                encode_hex(self.COLLATION_ADDED_TOPIC),
+                shard_id_topic_hex,
+            ],
+        )
+        for log in new_logs:
+            yield parse_collation_added_log(log)
+
+    def get_next_log(self):
+        new_logs = self._get_new_logs()
+        self.new_logs.extend(new_logs)
+        if len(self.new_logs) == 0:
+            raise NextLogUnavailable("No more next logs")
+        return self.new_logs.pop()
+
+    # TODO: this method may return wrong result when new logs arrive before the logs inside
+    #       `self.new_logs` are consumed entirely. This issue can be resolved by saving the
+    #       status of `new_logs`, `unchecked_logs`, and `current_score`, when it start to run
+    #       `GUESS_HEAD`. If there is a new block arriving, just restore them to the saved status,
+    #       append new logs to `new_logs`, and re-run `GUESS_HEAD`
+    def fetch_candidate_head(self):
+        # Try to return a log that has the score that we are checking for,
+        # checking in order of oldest to most recent.
+        unchecked_logs = pipe(
+            self.unchecked_logs,
+            enumerate,
+            tuple,
+            reversed,
+            tuple,
+        )
+        for idx, log_entry in unchecked_logs:
+            if log_entry['score'] == self.current_score:
+                return self.unchecked_logs.pop(idx)
+        # If no further recorded but unchecked logs exist, go to the next
+        # is_new_head = true log
+        while True:
+            try:
+                log_entry = self.get_next_log()
+            # TODO: currently just raise when there is no log anymore
+            except NextLogUnavailable:
+                # TODO: should returns the genesis collation instead or just leave it?
+                raise NoCandidateHead("No candidate head available")
+            if log_entry['is_new_head']:
+                break
+            self.unchecked_logs.append(log_entry)
+        self.current_score = log_entry['score']
+        return log_entry
+
+    def clean_logs(self):
+        self.new_logs = []
+        self.unchecked_logs = []
+
+    def add_log(self, log_entry):
+        collation_height = log_entry['score']
+        while len(self.headers_by_height) <= collation_height:
+            self.headers_by_height.append([])
+        self.headers_by_height[collation_height].append(log_entry['header'])
```

Comment:
Sorry I didn't get your meaning. In `add_log`, a log is add to the end of the list of its height. So within the list of a specific height in `headers_by_height`, later logs are in the behind of the list. And every time `candidate_heads` is called, all logs happened since last time `candidate_heads` is called are handled by `add_logs`, in the order from older to newer, ordering by period, ascending? or am I understanding wrong?